pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -353,15 +354,37 @@ public void reloadSegment(String segmentName, IndexLoadingConfig indexLoadingCon
     indexLoadingConfig.setSegmentTier(segmentTier);
     indexLoadingConfig.setTableDataDir(_tableDataDir);
     indexLoadingConfig.setInstanceTierConfigs(_tableDataManagerConfig.getInstanceTierConfigs());
+    indexLoadingConfig.setSegmentTier(zkMetadata.getTier());
     File indexDir = getSegmentDataDir(segmentName, segmentTier, indexLoadingConfig.getTableConfig());
     try {
+
+      boolean shouldDownload = forceDownload || !hasSameCRC(zkMetadata, localMetadata);
+
+      if (!shouldDownload) {
+        // We should first try to reuse directory
+        Properties loaderContextProps = new Properties();
+        loaderContextProps.put(CommonConstants.Server.DIRECTORY_LOADER_PURPOSE_CONFIG,
+            CommonConstants.Server.DIRECTORY_LOADER_PURPOSE_RELOAD);
+        SegmentDirectory segmentDirectory =
+            initSegmentDirectory(segmentName, String.valueOf(zkMetadata.getCrc()), indexLoadingConfig,
+                loaderContextProps);
+        boolean needReprocess = ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema);
+        if (!needReprocess) {
+          // No reprocessing needed, reuse the same segment
+          ImmutableSegment segment = ImmutableSegmentLoader.load(segmentDirectory, indexLoadingConfig, schema);
+          addSegment(segment);
+          return;
+        } else {
```

Comment:
looks like, if it takes the else-branch, the SegmentDirectory object may be created again at L402. 

I wonder if the changes can be simplified by checking `needPreprocess` after segmentDirectory is created at L402. If no need to preprocess, just add segment and done.

Reply:
That segmentdirectory is created with a backup dir. Which only supports copyTo(File dest) and not other functionalities..